### ISSUE TITLE
[BUG?] npm link causes Webpack to look for babel-plugin in linked package node_modules.

### ISSUE THREAD

**Issue by trusktr** [webpack/webpack#1866]
I've linked a package that is outside of the project by first running `npm link` in the outside package, then `npm link name-of-package` inside the project.

I am now getting this error:

```
Module build failed: ReferenceError: Unknown plugin "transform-es2015-arrow-functions" specified in "base" at 0, attempted to resolve relative to "/path/to/location/far/away/from/project/name-of-package"
```

Of course there's no babel plugin there.

**Comment by trusktr** [comment]
Is there an intention for package authors to depend on certain babel plugins? Maybe that's why this error happens, since the package is outside my project the babel plugin wasn't found there? If that's the case, this would be a feature like how browserify lets developers specify transforms, and wouldn't be a bug, but would need to be documented for library authors to take advantage of.

Nonetheless, it should still work if my project has the needed plugins installed.

**Comment by gileze33** [comment]
Just ran into the same issue, did you manage to find a work around? `npm link` is obviously super important for our workflow

**Comment by trusktr** [comment]
@gileze33 Unfortunately I haven't found a workaround. What I'm doing now is just `git push`ing my package to GitHub, then re-installing the new one each time I push, for development. It's a slower workflow, but at least it works for now. `package.json` just has the dependency listed something like:

```
{
  "dependencies": {
    "package-name": "username/package-name#master"
  }
}
```

**Comment by basham** [comment]
I'm avoiding the issue by sticking with `babel-core@6.1.21`, the latest version that works. I've been able to upgrade other babel loaders, plugins, and presets without issue in the meantime. Just going to wait for a solution.

**Comment by trusktr** [comment]
Ah, good to know! I'll try giving that a shot too.

/#!/JoePea

**Comment by trusktr** [comment]
@basham Indeed, `babel-core@6.1.21` works! I'm using that for now, as `npm link`ing makes dev life a lot easier.

**Comment by trusktr** [comment]
Oh, wait, so is this an issue with Babel, not Webpack?

**Comment by basham** [comment]
I highly suspect it is due to this [`6.2.0` change](https://github.com/babel/babel/blob/master/CHANGELOG.md#620):

> Infer filename from the base directory when resolving plugins and presets.

The issue seems to be that babel plugins now want to be relative to their module, not relative to the main project. Webpack isn't aware of Babel plugins to resolve it like it does with the [`resolve`](https://webpack.github.io/docs/configuration.html#resolve) and [`resolveLoader`](https://webpack.github.io/docs/configuration.html#resolveloader) configurations. Webpack could potentially solve this, but I'm not convinced it should be Webpack's responsibility to accommodate how babel does its resolving.

**Comment by kintel** [comment]
This sounds a bit like webpack isn't correctly excluding node_modules. Perhaps there's a symlink resolution issue causing it to run the babel loader on npm linked modules, even though the links are located in node_modules.
Browserify had [a very similar bug](https://github.com/substack/node-browserify/issues/1386) a while ago, and I experience this bug to be of similar nature.

**Comment by kintel** [comment]
I investigated a bit more and I can reproduce this problem without any usage of babel.
A simple npm linked module and using the brfs transform loader triggers the same issue:
```
    module: {
      loaders: [ { loader: "transform?brfs" } ]
    }
```

I don't currently have any good workaround for this, except ensuring that you only npm link fully transformed modules. Also, browserify seems to handle all this pretty well (although that's not a good workaround seen from the perspective of this project :/ )

**Comment by slightlytyler** [comment]
Also experiencing this issue. @basham your suggestion works but doesn't truly solve the problem, just circumvents the issue of webpack loading linked modules from their source directory for the case of transpiling with babel. In my case the babel transform error is gone but my linter (loaded through webpack) is still printing errors for the linked module.

**Comment by zzarcon** [comment]
I'm also having the same issue using `babel-plugin-transform-async-to-generator`, I just downgraded from `babel-core@6.7.2` to `babel-core@6.1.21` and it works...

Any update on this?

**Comment by trusktr** [comment]
Ah, the `require.resolve` trick makes it absolute, circumventing Webpack from looking for the plugin in the wrong place during build? Good workaround!

**Comment by kintel** [comment]
I believe plugins also suffer from this bug.
I'm using the ```ProvidePlugin``` to bind the three.js library to THREE:

```
    new webpack.ProvidePlugin({
      THREE: 'three'
    }),
```

This triggers the ```Module not found: Error: Cannot resolve module 'three' in ...``` error in my npm linked modules.

I did not manage to get the require.resolve trick to work for my babel setup. My workaround is to manually exclude regexps matching my linked modules.
The trick does work as a workaround for the provide plugin though.

**Comment by MrHayato** [comment]
I get a similar issue with typescript and ts-loader. It used to work on latest stable webpack 1, but upgrading to webpack 2, and updating the configuration to work with webpack 2, I get a ton of errors that say "Cannot find module", but if I `npm install` inside the npm linked folder, then the errors go away and everything works again...

**Comment by kaore** [comment]
+1

**Comment by devinivy** [comment]
I'm also having this issue when using `npm link`.

```
Module build failed: ReferenceError: Unknown plugin "transform-runtime" specified in "base" at 0, attempted to resolve relative to "/path/to/npm/linked/library"
```

**Comment by jamesknelson** [comment]
Just a note that the `require.resolve` workaround only works up until Webpack version 1.12.15. Later versions use `acorn` instead of `esprima` as the JS parser, and this seems to prevent parsing of the JSON in the query string somehow.

**Comment by HriBB** [comment]
> I'm avoiding the issue by sticking with babel-core@6.1.21, the latest version I've found that works with npm link. I've been able to upgrade other babel loaders, plugins, and presets without issue in the meantime. I also have the latest webpack and webpack-dev-server. Just going to wait for a solution.

Thanks a million @basham! I have been trying to get `npm link` to work with `webpack` and `babel` for some time now, and now it finally works with `babel-core@6.1.21`.

This is my webpack config if it helps anyone
```
const fs = require('fs')
const path = require('path')
const webpack = require('webpack')

const reactMdlSelectfieldPath = fs.realpathSync(
  path.resolve(__dirname, '..', 'node_modules', 'react-mdl-selectfield', 'src')
)

module.exports = {
  name: 'client',
  target: 'web',
  devtool: 'cheap-module-inline-source-map',
  entry: [
    'react-hot-loader/patch',
    'webpack-dev-server/client?http://localhost:3000',
    'webpack/hot/only-dev-server',
    path.resolve(__dirname, '..', 'client', 'index'),
  ],
  output: {
    path: path.resolve(__dirname, '..', 'dist'),
    filename: 'app.js',
    publicPath: '/static/',
  },
  resolve: {
    root: path.resolve(__dirname, '..', 'node_modules'),
    fallback: path.resolve(__dirname, '..', 'node_modules'),
    extensions: ['', '.js', '.scss', '.css'],
    alias: {
      config: path.resolve(__dirname, '..', 'config'),
      apollo: path.resolve(__dirname, '..', 'client', 'apollo'),
      components: path.resolve(__dirname, '..', 'client', 'components'),
      containers: path.resolve(__dirname, '..', 'client', 'containers'),
      reducers: path.resolve(__dirname, '..', 'client', 'reducers'),
      routes: path.resolve(__dirname, '..', 'client', 'routes'),
      store: path.resolve(__dirname, '..', 'client', 'store'),
      i18n: path.resolve(__dirname, '..', 'client', 'i18n'),
      'react-mdl-selectfield': reactMdlSelectfieldPath,
    }
  },
  resolveLoader: {
    root: path.resolve(__dirname, '..', 'node_modules'),
    fallback: path.resolve(__dirname, '..', 'node_modules'),
  },
  module: {
    loaders: [{
      test: /\.js$/,
      loader: 'babel',
      exclude: /(node_modules)/,
      include: [
        path.resolve(__dirname, '..', 'config'),
        path.resolve(__dirname, '..', 'client'),
        reactMdlSelectfieldPath,
      ],
    },{
      test: /\.(scss|css)$/,
      loader: 'style!css?sourceMap!sass?sourceMap!postcss',
    },{
      test: /\.(png|jpg|svg)$/,
      loader: 'url-loader?limit=8192',
    },{
      test: /masonry|imagesloaded|fizzy\-ui\-utils|desandro\-|outlayer|get\-size|doc\-ready|eventie|eventemitter/,
      loader: 'imports?define=>false&this=>window',
    }]
  },
  sassLoader: {
    includePaths: [path.resolve(__dirname, '..', 'client', 'styles')],
  },
  postcss: function() {
    return [
      require('autoprefixer'),
    ]
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env': {
        NODE_ENV: '"development"'
      },
      __DEV__: true,
      __PROD__: false,
    }),
    new webpack.ContextReplacementPlugin(/moment[\/\\]locale$/, /sl/),
    new webpack.HotModuleReplacementPlugin(),
  ]
}
```

**Comment by paulpooch** [comment]
If you're here you're probably looking for this solution (at least I was):

http://webpack.github.io/docs/troubleshooting.html#npm-linked-modules-doesn-t-find-their-dependencies

**Comment by russellrobinson** [comment]
Here's my workaround for linking problems:
Project **proj**, uses module **mod**, which has a dependency on **dep**
Working in folder /home/src

1. Install **proj** and **mod** folders in /home/src: /home/src/proj, /home/src/mod

**Comment by Agamennon** [comment]
try to add  ` resolve: {
        symlinks: false
    },` to your webpack.config

**Comment by f15gdsy** [comment]
@Agamennon Thank you, it works great!

**Comment by everystone** [comment]
@Agamennon Thanks so much, I spent all day looking for that.

**Comment by dbarabander** [comment]
I tried everything here, only switching my node version from `6.9.2` to `6.0.0` fixed my issue...

**Comment by p4bloch** [comment]
Had a special use case which wasn't solved by adding `symlinks: false`, and even when I added an exclude rule that made it work, IE10 in a VM wasn't able to render the page.

Since this is the main issue for this kind of problems (at least the first one I landed to), I wanted to share that as suggested by @ganmor [here](https://github.com/webpack/webpack/issues/985#issuecomment-256300077), using https://github.com/wix/wml and avoiding symlinks entirely solved all my issues immediately.

**Comment by joaoreynolds** [comment]
`{symlinks: false}` fixes the error, however it disrupts webpack's watch function (and therefore webpack-dev-server will not recompile when changing files in the "linked" dependency).

Anybody found another solution?

**Comment by gregorskii** [comment]
Symlink false does not work with Gatsby which uses webpack 1.

**Comment by wmzy** [comment]
@joaoreynolds 
Let the `followSymlinks` to be `true` work for me : [node_modules/watchpack/lib/DirectoryWatcher.js#L52](https://github.com/webpack/watchpack/blob/9fc9be000ebd3377418c33be21764867e7a5cbcb/lib/DirectoryWatcher.js#L52)

But why not webpack turn on the `followSymlinks` option when `symlinks` is `false`?

**Comment by txinfo** [comment]
I found the same as @wmzy after my own lengthy investigation. I have opened an issue on watchpack's issue tracker to make that option configurable: https://github.com/webpack/watchpack/issues/61

**Comment by nreoch25** [comment]
Something I did which isn't the most ideal is just install the babel plugin as a devDependency in the module you are linking into your project.  The babel-plugin-dynamic-import-node was the culprit for me so I just installed that. Errors cleared up and I can continue working on my package.  Pretty much a time saver so I can stop looking into this issue and move forward.

**Comment by mbardelmeijer** [comment]
Any updates on this issue? 😄

**Comment by birdy-** [comment]
Just opened a PR on watchpack:
https://github.com/webpack/watchpack/pull/72

**Comment by birdy-** [comment]
Add an issue with possible solution feel free to comment or thumbs up this future change.
https://github.com/webpack/webpack/issues/6845

**Comment by JamesRamm** [comment]
My workaround (for webpack 3) which is compatible with webpack-dev-server:

First set the `resolve.modules` property:

```js
resolve: { modules: [path.resolve(__dirname, 'node_modules'] }
```

It is important that you supply the abs path to your `node_modules`

The, for each babel preset and plugin, explicitly use `require.resolve`.

```js
query: {
    plugins: [ require.resolve('babel-plugin-transform-object-rest-spread')],
   presets: [ require.resolve('babel-preset-react') ..etc.. ]
}
```

Note this means you need to explicitly move any babel setup you have in a babelrc file into your webpack config (in my case it was moving the plugins defs)

This compiles and when I make a change in a `yarn link`'d module, dev-server is picking up the changes

**Comment by alexander-akait** [comment]
I think we can close it, nothing to fix on webpack side, `require.resolve` is good solution, also can be relevant tot his https://classic.yarnpkg.com/blog/2018/02/15/nohoist/

**Comment by nico-ledesma** [comment]
It's May 2022 and this issue was still present for me: `babel-loader` failed when processing symlinked packages added via `npm link`.

Hope this helps!

**TL;DR**
Add this to `webpack.config.js` when watching:
```javascript
resolve.alias = {
    'core-js' : path.resolve( __dirname, './node_modules/core-js' ),
    'regenerator-runtime' : path.resolve( __dirname, './node_modules/regenerator-runtime' )
} 
```

**Long explanation**
- I was using `npm link` in my main project to add symlinked dependencies to other modules.
- These modules lived together in Lerna monorrepo, initialized by running `lerna bootstrap`.
- `babel-loader` failed because it was looking for `core-js` and `regenerator-runtime` in each symlinked modules `node_modules` dir.
- I was using `webpack@5`, `@babel/core@7` `@babel/preset-env@7` like this:
```javascript
[ "@babel/preset-env", {
    "useBuiltIns": "usage",
    "corejs" : 3
} ]
```

Solutions that were not working:
- Using absolute paths for presets, i.e. `require.resolve( @babel/preset-env)`.
- Using `resolve.symlinks : false` in Webpack:
  - This fixes `babel-loader` problem, but Webpack won't recompile if the symlinked modules change.
  - Using `watchOptions.followSymlinks : true` doesn't fix this Webpack watch problem either.

If you can't avoid using `babel-loader` when watching, defining those aliases does the trick.

**Comment by alexander-akait** [comment]
@nico-ledesma Can you open a new issue and describe it, we will investigate it, thanks

**Comment by toFrankie** [comment]
👀
